sys/net/routing/nhdp/iib_table.c

```diff
@@ -0,0 +1,698 @@
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     nhdp
+ * @{
+ *
+ * @file
+ * @brief       Interface Information Base implementation for NHDP
+ *
+ * @author      Fabian Nack <nack@inf.fu-berlin.de>
+ *
+ * @}
+ */
+
+#include "mutex.h"
+#include "timex.h"
+#include "vtimer.h"
+#include "utlist.h"
+#include "kernel_types.h"
+
+#include "rfc5444/rfc5444_iana.h"
+#include "rfc5444/rfc5444_writer.h"
+
+#include "iib_table.h"
+#include "nhdp_address.h"
+#include "nhdp_writer.h"
+
+/* Internal variables */
+static mutex_t mtx_iib_access = MUTEX_INIT;
+static iib_base_entry_t *iib_base_entry_head = NULL;
+
+/* Internal function prototypes */
+static void rem_link_set_entry(iib_base_entry_t *base_entry, iib_link_set_entry_t *ls_entry);
+static void cleanup_link_sets(nhdp_addr_entry_t *rem_list);
+static iib_link_set_entry_t* add_default_link_set_entry(iib_base_entry_t *base_entry, timex_t *now,
+        uint64_t val_time);
+static void reset_link_set_entry(iib_link_set_entry_t *ls_entry, timex_t *now, uint64_t val_time);
+static iib_link_set_entry_t* update_link_set(iib_base_entry_t *base_entry, nib_entry_t *nb_elt,
+        nhdp_addr_entry_t *send_list, timex_t *now, uint64_t val_time, uint8_t sym, uint8_t lost);
+static void release_link_tuple_addresses(iib_link_set_entry_t *ls_entry);
+
+static int update_two_hop_set(iib_base_entry_t *base_entry, iib_link_set_entry_t *ls_entry,
+        nhdp_addr_entry_t *th_sym_list, nhdp_addr_entry_t *th_rem_list,
+        timex_t *now, uint64_t val_time);
+static uint8_t rem_exst_th_entries(iib_base_entry_t *base_entry,
+        iib_two_hop_set_entry_t *ts_elt, nhdp_addr_entry_t *th_sym_list);
+static uint8_t rem_non_sym_th_entries(iib_base_entry_t *base_entry,
+        iib_two_hop_set_entry_t *ts_elt, nhdp_addr_entry_t *th_rem_list);
+static int add_two_hop_entry(iib_base_entry_t *base_entry, iib_link_set_entry_t *ls_entry,
+        nhdp_addr_t *th_addr, timex_t *now, uint64_t val_time);
+static void rem_two_hop_entry(iib_base_entry_t *base_entry, iib_two_hop_set_entry_t *th_entry);
+
+static void wr_update_ls_status(iib_base_entry_t *base_entry,
+        iib_link_set_entry_t *ls_elt, timex_t *now);
+static void sec_thirteen_one(iib_link_set_entry_t *ls_entry);
+static void sec_thirteen_two(iib_base_entry_t *base_entry,
+        iib_link_set_entry_t *ls_entry, timex_t *now);
+static void sec_thirteen_three(iib_link_set_entry_t *ls_entry, timex_t *now);
+
+static timex_t get_max_timex(timex_t time_one, timex_t time_two);
+static iib_link_tuple_status_t get_tuple_status(iib_link_set_entry_t *ls_entry, timex_t *now);
+
+
+/*---------------------------------------------------------------------------*
+ *                       Interface Information Base API                      *
+ *---------------------------------------------------------------------------*/
+
+int iib_register_if(kernel_pid_t pid)
+{
+    iib_base_entry_t *new_entry = (iib_base_entry_t*) malloc(sizeof(iib_base_entry_t));
```

Comment:
Wouldn't it be possible to declare a static array of a sensible maximum size here instead?